pkg/contexts/ocm/blobhandler/handlers/generic/npmjs/publish.go

```diff
@@ -0,0 +1,214 @@
+	hash.Write(data)
+	return "sha512-" + base64.StdEncoding.EncodeToString(hash.Sum(nil))
+}
+
+func createSha1(data []byte) string {
+	hash := sha1.New() //nolint:gosec // older npm (prior to v5) uses sha1
+	hash.Write(data)
+	return hex.EncodeToString(hash.Sum(nil))
+}
+
+// Read package.json and README.md from tarball to create Package object.
+func prepare(data []byte) (*Package, error) {
+	gz, err := gzip.NewReader(bytes.NewReader(data))
+	if err != nil {
+		return nil, err
+	}
+	tr := tar.NewReader(gz)
+	var (
+		pkgData []byte
+		readme  []byte
+	)
+	for {
+		thr, e := tr.Next()
+		if e != nil {
+			if errors.Is(e, io.EOF) {
+				break
+			}
+			return nil, e
+		}
+		if pkgData != nil && readme != nil {
+			break
+		}
+		switch thr.Name {
+		case "package/package.json":
+			pkgData, err = io.ReadAll(tr)
+			if err != nil {
+				return nil, fmt.Errorf("read package.json failed, %w", err)
+			}
+		case "package/README.md":
+			readme, err = io.ReadAll(tr)
+			if err != nil {
+				return nil, fmt.Errorf("read README.md failed, %w", err)
+			}
+		}
+	}
+
+	// fetch some information from package.json
+	if len(pkgData) == 0 {
+		return nil, fmt.Errorf("package.json is empty")
+	}
+	var pkgJson map[string]interface{}
+	err = json.Unmarshal(pkgData, &pkgJson)
+	if err != nil {
+		return nil, fmt.Errorf("read package.json failed, %w", err)
+	}
+	if pkgJson["name"] == "" {
+		return nil, fmt.Errorf("package.json's name is empty")
+	}
+	if pkgJson["version"] == "" {
+		return nil, fmt.Errorf("package.json's version is empty")
+	}
+
+	// create package object
+	var pkg Package
+	pkg.Name = pkgJson["name"].(string)
+	pkg.Version = pkgJson["version"].(string)
+	pkg.Description = pkgJson["description"].(string)
+	pkg.Readme = string(readme)
+	pkg.Dist.Shasum = createSha1(data)
+	pkg.Dist.Integrity = createSha512(data)
+	return &pkg, nil
+}
+
+// Check if package already exists in npmjs registry. If it does, checks if it's the same.
+func packageExists(repoUrl string, pkg Package, token string) (bool, error) {
+	client := http.Client{}
+	req, err := http.NewRequestWithContext(context.Background(), http.MethodGet, repoUrl+"/"+url.PathEscape(pkg.Name)+"/"+url.PathEscape(pkg.Version), nil)
+	if err != nil {
+		return false, err
+	}
+	req.Header.Set("authorization", "Bearer "+token)
+	resp, err := client.Do(req)
+	if err != nil {
+		return false, err
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode == http.StatusNotFound {
+		// artifact doesn't exist, it's safe to upload
+		return false, nil
+	}
+
+	// artifact exists, let's check if it's the same
+	all, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return false, err
+	}
+	if resp.StatusCode != http.StatusOK {
+		return false, fmt.Errorf("http (%d) - %s", resp.StatusCode, string(all))
+	}
+	var data map[string]interface{}
+	err = json.Unmarshal(all, &data)
+	if err != nil {
+		return false, err
+	}
+	dist := data["dist"].(map[string]interface{})
+	if pkg.Dist.Integrity == dist["integrity"] {
```

Comment:
Is it always the same digest type, or should we check the type analysing the digest prefix (<type>:<hash>)

Reply:
as far as I've seen it, it's always the same
https://docs.npmjs.com/cli/v6/configuring-npm/package-lock-json#integrity

when SHA512 isn't sufficient anymore for NPM, then we can adapt the coding